syntax = "proto3";

import "db_base.proto";

option go_package = "pb/deploy";

option cc_generic_services = true;

message IpPort
{
  string ip = 1;
  uint32 port = 2;
}

message database_server_db
{
  option(OptionTableName)  = "database_server_db";
  option(OptionPrimaryKey) = "id";
  option(OptionAutoIncrementKey) = "id";
  uint32 id = 1;
  string ip = 2;
  uint32 port = 3;
  string db_host = 4;
  string db_user = 5;
  string db_password = 6;
  uint32 db_port = 7;
  string db_dbname = 8;
  uint32 lobby_id = 9;
}

message servers_info_data
{
  database_server_db database_info = 1;
  login_server_db login_info = 2;
  controller_server_db controller_info = 3;
  gate_server_db gate_info = 4;
  redis_server_db redis_info = 5;
  lobby_server_db lobby_info = 6;
}

message login_server_db
{
  option(OptionTableName)  = "login_server_db";
  option(OptionPrimaryKey) = "id";
  option(OptionAutoIncrementKey) = "id";
  uint32 id = 1;
  string ip = 2;
  uint32 port = 3;
  uint32 zone_id = 4;
}

message controller_server_db
{
  option(OptionTableName)  = "controller_server_db";
  option(OptionPrimaryKey) = "id";
  option(OptionAutoIncrementKey) = "id";
  uint32 id = 1;
  string ip = 2;
  uint32 port = 3;
  uint32 lobby_id = 4;
}

message lobby_server_db
{
  option(OptionTableName)  = "lobby_server_db";
  option(OptionPrimaryKey) = "id";
  option(OptionAutoIncrementKey) = "id";
  uint32 id = 1;
  string ip = 2;
  uint32 port = 3;
}

message redis_server_db
{
  option(OptionTableName)  = "redis_server_db";
  option(OptionPrimaryKey) = "id";
  option(OptionAutoIncrementKey) = "id";
  uint32 id = 1;
  string ip = 2;
  uint32 port = 3;
  uint32 zone_id = 4;
}

message game_server_db
{
  option(OptionTableName)  = "game_server_db";
  option(OptionPrimaryKey) = "id";
  option(OptionAutoIncrementKey) = "id";
  uint32 id = 1;
  string ip = 2;
  uint32 port = 3;
  uint32 lobby_id = 4;
}

message gate_server_db
{
  option(OptionTableName)  = "gate_server_db";
  option(OptionPrimaryKey) = "id";
  option(OptionAutoIncrementKey) = "id";
  uint32 id = 1;
  string ip = 2;
  uint32 port = 3;
  uint32 zone_id = 4;
}


message NodeInfoRequest
{
  uint32 zone_id = 1;
  uint32 group_id = 2;
}

message ControllerNodesDb
{
  repeated controller_server_db controllers = 1;
}

message LoginNodesDb
{
  repeated login_server_db login_nodes = 1;
}

message NodeInfoResponse
{
  servers_info_data info = 1;
}

message StartGsRequest
{
  uint32 zone_id = 1;
  game_server_db my_info = 2;
  IpPort rpc_client = 3; 
}

message StartGsResponse
{
  game_server_db my_info = 1;
  uint32 error_no = 2;
  redis_server_db redis_info = 5;
}

message LobbyServerRequest
{
  uint32 lobby_id = 1;
}

message LobbyServerResponse
{
  lobby_server_db info = 1;
}

message LobbyInfoResponse
{
  ControllerNodesDb lobby_controllers = 1;
}

message GsNodeIdRequest
{
}

message GsNodeIdResponse
{
  uint32 node_id = 1;
}

service DeployService {
  rpc GetNodeInfo (NodeInfoRequest) returns (NodeInfoResponse);
  rpc StartGs (StartGsRequest) returns (StartGsResponse);
  rpc GetGsNodeId (GsNodeIdRequest) returns (GsNodeIdResponse);
}
